From: Harsh Prateek Bora <harsh@linux.vnet.ibm.com>
Subject: [PATCH 074/111]     fs/9p: mkdir fix for setting S_ISGID bit as per parent directory
Git-commit: 7c7298cffc8f4417c95117c2a432f962c066499d
Patch-Mainline: 2.6.37
References: FATE#311639

    The current implementation of 9p client mkdir function does not
    set the S_ISGID mode bit for the directory being created if the
    parent directory has this bit set. This patch fixes this problem
    so that the newly created directory inherits the gid from parent
    directory and not from the process creating this directory, when
    the S_ISGID bit is set in parent directory.

	upstream commit 7c7298cffc8f4417c95117c2a432f962c066499d

    Signed-off-by: Harsh Prateek Bora <harsh@linux.vnet.ibm.com>
    Signed-off-by: Eric Van Hensbergen <ericvh@gmail.com>
Acked-by: Bruce Rogers <brogers@suse.com>
---
 fs/9p/vfs_inode.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/fs/9p/vfs_inode.c b/fs/9p/vfs_inode.c
index 3ebbe21..428111f 100644
--- a/fs/9p/vfs_inode.c
+++ b/fs/9p/vfs_inode.c
@@ -873,6 +873,8 @@ static int v9fs_vfs_mkdir_dotl(struct inode *dir, struct dentry *dentry,
 	v9ses = v9fs_inode2v9ses(dir);
 
 	mode |= S_IFDIR;
+	if (dir->i_mode & S_ISGID)
+		mode |= S_ISGID;
 	dir_dentry = v9fs_dentry_from_dir_inode(dir);
 	dfid = v9fs_fid_lookup(dir_dentry);
 	if (IS_ERR(dfid)) {
-- 
1.7.1.1

